// @homebridge/ciao/lib/coder/records/UnsupportedRecord.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UnsupportedRecord = void 0;
const ResourceRecord_1 = require("../ResourceRecord");
class UnsupportedRecord extends ResourceRecord_1.ResourceRecord {
    constructor(header, data) {
        super(header);
        this.data = data;
    }
    encodeRData(coder, buffer, offset) {
        return this.data.copy(buffer, offset);
    }
    getRDataEncodingLength() {
        return this.data.length;
    }
    clone() {
        return new UnsupportedRecord(this.getRecordRepresentation(), this.data);
    }
    dataAsString() {
        return this.data.toString("base64");
    }
    dataEquals(record) {
        return this.data.toString("base64") === record.data.toString("base64");
    }
    static decodeData(coder, header, buffer, offset) {
        const data = buffer.slice(offset);
        return {
            data: new UnsupportedRecord(header, data),
            readBytes: data.length,
        };
    }
}
exports.UnsupportedRecord = UnsupportedRecord;
